tools/cell_census_builder/validate.py

```diff
@@ -276,59 +276,86 @@ def _validate_X_layers_contents_by_dataset(args: Tuple[str, str, Dataset, List[E
-                    actual_census_gene_count = _count_elements(
-                        exp.ms["RNA"][FEATURE_DATASET_PRESENCE_MATRIX_NAME], [dataset.soma_joinid]
-                    )
-
-                    # mapping the var_soma_joinid to the anndata feature_ids
-                    actual_genes = query.var(column_names=["soma_joinid", "feature_id"]).concat().to_pandas()
-                    gene_mapping = ad.var.join(actual_genes.set_index("feature_id"))[["soma_joinid"]]
-                    gene_mapping["ad_id"] = range(gene_mapping.shape[0])
-                    gene_mapping = gene_mapping.set_index("soma_joinid")
-                    actual_x = (
-                        query.X("raw")
-                        .tables()
-                        .concat()
-                        .to_pandas()
-                        .join(gene_mapping, on="soma_dim_1")[["soma_dim_0", "ad_id", "soma_data"]]
-                    )
-                    # translating the rows
-                    actual_x.soma_dim_0 = actual_x.soma_dim_0 - actual_x.soma_dim_0.min()
-                    # convert to sparse matrix
-                    actual_x_coo = sparse.coo_matrix(
-                        (actual_x.soma_data.array, (actual_x.soma_dim_0.array, actual_x.ad_id.array))
-                    )
-            if actual_x_coo is not None:
-                expected_ad_x_coo = expected_ad_x.tocoo()
-                assert expected_ad_x_coo.nnz == actual_x_coo.nnz, error.format(
-                    "nnz", actual_x_coo.nnz, expected_ad_x.nnz
-                )
-                assert expected_ad_x_coo.shape == actual_x_coo.shape, error.format(
-                    "shape", actual_x_coo.shape, expected_ad_x_coo.shape
+                expected_ad_x, expected_ad_var = ad.raw.X, ad.raw.var
+
+            # get the joinids for the obs axis
+            obs_joinids = (
+                exp.obs.read(
+                    column_names=["soma_joinid", "dataset_id"], value_filter=f"dataset_id == '{dataset.dataset_id}'"
                 )
-                assert (
-                    expected_ad_x_coo != actual_x_coo
-                ).nnz == 0, f"{eb.name}:{dataset.dataset_id} the X matrix elements are not equal."
-            expected_gene_count = np.count_nonzero(expected_ad_x.sum(axis=0))
-            assert expected_gene_count == actual_census_gene_count, error.format(
-                "gene_count", actual_census_gene_count, expected_gene_count
+                .concat()
+                .to_pandas()
+            )
+
+            assert ad.n_obs == len(obs_joinids)
+            if len(obs_joinids) == 0:
+                continue
+
+            X_raw = (
+                exp.ms[MEASUREMENT_RNA_NAME]
+                .X["raw"]
+                .read((obs_joinids.soma_joinid.to_numpy(), slice(None)))
+                .tables()
+                .concat()
+            )
+            data = X_raw["soma_data"].to_numpy()
```

Comment:
Isn't it expensive to expand out the whole sparse matrix?

Reply:
sure, but if you want to compare every value withe AnnData, you don't have any choice.  You must read it all in order to check for equality.